### PR TITLE
Updated API schema, SDK endpoint now uses feathersjs

### PR DESCRIPTION
**Pull request by PatrickShaw**
Implementing multiple tasks/features in a single commit 🥇
image

Updated the API schema. You now need `HasId<...>` wrapped around your object if you know it's going to have an ID. Also made the `/generatesdk` endpoint a REST compliant end point. Also improved `BuildStatusChip`'s performance slightly (was originally creating a totally new `Styled` component whenever it was rendered); because of this, I also swapped out mui-next's `withStyles` method for the `react-jss` equivalent (they're inter-operable, to knowledge, and I needed to pass props into different style classes.

Oh and I changed `model/Specification` to `model/Spec` cause it's getting too tiring to type out all the time.

### CODE REVIEW

File: src/client/BackendClient.tsx [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import feathers from '@feathersjs/client';
 import io from 'socket.io-client';
+
```
Comment (darvid7) [inline src/client/BackendClient.tsx]: lgtm

File: src/client/sdkGeneration.tsx [hunk 2]
```diff
@@ -19,10 +18,9 @@ export async function generateSdk(spec: Specification): Promise<any> {
     body: JSON.stringify(body),
     headers: { 'Content-Type': 'application/json' }
   });
-  const fulfilled = await response.json();
-  console.log(fulfilled);
-  if (fulfilled.type == 'error') {
-    return BAD_SPECIFICATION;
+  const json = await response.json();
+  if (json.type === 'error') {
+    throw new Error(json.message);
```
Comment (darvid7) [inline src/client/sdkGeneration.tsx]: are we sure we want to throw? will this crash the backend?

File: src/model/Spec.tsx [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { HasId } from 'model/Entity';
+import { Sdk } from 'model/Sdk';
+export interface Spec {
```
Comment (darvid7) [inline src/model/Spec.tsx]: Should we add options and languages?

File: src/backend/index.tsx [hunk 4]
```diff
@@ -39,39 +60,29 @@ async function run(port: number) {
       })
     )
     .get('/', (req, res) => res.redirect('/docs'))
-    .use('/specifications', specifications)
+    .use('/specifications', specs)
```
Comment (darvid7) [inline src/backend/index.tsx]: magic crud :D 👍

File: src/client/sdkGeneration.tsx (hunk 2 above)
Comment (PatrickShaw) [reply src/client/sdkGeneration.tsx]: No, it returns a error message to the user with HTTP 500 Internal Server Error. Note that I don't think it should return 500 but that's better than what it was doing before (returning 200, to memory).

File: src/model/Spec.tsx (hunk 3 above)
Comment (PatrickShaw) [reply src/model/Spec.tsx]: One day :3, in another PR

File: src/view/basic/BuildStatusChip.tsx [hunk 5]
```diff
@@ -1,14 +1,39 @@
...
+    case BuildStatus.NOTRUN:
+      return grey[300];
+    case BuildStatus.RUNNING:
+      return amber[500];
+    case BuildStatus.SUCCESS:
+      return green[600];
+    case BuildStatus.FAIL:
+      return red[900];
+  }
+};
```
Comment (PatrickShaw) [inline src/view/basic/BuildStatusChip.tsx]: Obviously a less verbose way to do this but i think we should worry bout it later.

File: src/model/Plan.tsx [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+export enum BuildStatus {
+  NOTRUN,
+  RUNNING,
+  SUCCESS,
+  FAIL
+}
+
+export interface SdkResponse {}
```
Comment (hferrier) [inline src/model/Plan.tsx]: `ERROR: src/model/Plan.tsx[18, 1]: An empty interface is equivalent to {}`
That's what tslint would say if it was enabled in this branch.
Why is this lil guy here? What's he doing? Where's he going? What is his purpose?
Doesn't appear to be used anywhere yet, could it maybe just be removed for now?

Comment (PatrickShaw) [reply src/model/Plan.tsx]: Annddd the answer iissssss... No idea why he's there. Thank you :)

**Comment by PatrickShaw** [comment]
@hferrier 
I'm willing to resolve all the merge conflicts :P

**Pull request merged**